File: lib/data/models/pomodoro_task_reportage_model.dart
import 'package:sptimer/data/enums/task_status.dart';

const _kTaskReportageId = 'kTaskReportageId';
const _kPomodoroTaskId = 'kPomodoroTaskId';
const _kStartDate = 'kStartDate';
const _kEndDate = 'kEndDate';
const _kTaskName = 'kTaskName';
const _kStatus = 'kStatus';

class PomodoroTaskReportageModel {
  PomodoroTaskReportageModel({
    this.id,
    this.endDate,
    required this.startDate,
    required this.taskStatus,
    required this.pomodoroTaskId,
    required this.taskName,
  });
  final int? id;
  final DateTime startDate;
  final DateTime? endDate;
  final TaskStatus taskStatus;
  final int pomodoroTaskId;
  final String taskName;

  PomodoroTaskReportageModel copyWith({
    int? id,
    int? pomodoroTaskId,
    DateTime? startDate,
    DateTime? endDate,
    String? taskName,
    TaskStatus? taskStatus,
  }) {
    return PomodoroTaskReportageModel(
      id: id ?? this.id,
      pomodoroTaskId: pomodoroTaskId ?? this.pomodoroTaskId,
      startDate: startDate ?? this.startDate,
      endDate: endDate ?? this.endDate,
      taskName: taskName ?? this.taskName,
      taskStatus: taskStatus ?? this.taskStatus,
    );
  }

  Map<String, dynamic> toMap() {
    return {
      _kTaskReportageId: id,
      _kPomodoroTaskId: pomodoroTaskId,
      _kStartDate: startDate.toString(),
      _kEndDate: endDate?.toString(),
      _kTaskName: taskName,
      _kStatus: taskStatus.index,
    };
  }

  factory PomodoroTaskReportageModel.fromMap(Map<dynamic, dynamic> map) {
    return PomodoroTaskReportageModel(
      id: map[_kTaskReportageId],
      pomodoroTaskId: map[_kPomodoroTaskId],
      startDate: DateTime.parse(map[_kStartDate]),
      endDate: map[_kEndDate] != null ? DateTime.parse(map[_kEndDate]) : null,
      taskName: map[_kTaskName],
      taskStatus: TaskStatus.values[map[_kStatus]],
    );
  }
}

File: android/app/src/main/java/com/example/pomotimer/models/PomodoroTaskReportageModel.java
package com.example.pomotimer.models;

import com.example.pomotimer.Constants;
import java.util.HashMap;
import java.util.Map;



public class PomodoroTaskReportageModel {
    public PomodoroTaskReportageModel(
            Integer id,
            String startDate,
            String endDate,
            int taskStatus,
            int pomodoroTaskId,
            String taskName

    ) {
        this.id = id;
        this.startDate = startDate;
        this.endDate = endDate;
        this.taskStatus = taskStatus;
        this.pomodoroTaskId = pomodoroTaskId;
        this.taskName = taskName;
    }

    public Integer id;
    public String startDate;
    public String endDate;
    public int taskStatus;
    public int pomodoroTaskId;
    public String taskName;


    public Map<String, Object> toMap() {
        Map<String, Object> map = new HashMap<>();
        map.put(Constants.kTaskReportageId, id);
        map.put(Constants.kPomodoroTaskId, pomodoroTaskId);
        map.put(Constants.kStartDate, startDate);
        map.put(Constants.kEndDate, endDate);
        map.put(Constants.kTaskName, taskName);
        map.put(Constants.kStatus, taskStatus);
        return map;
    }

    public static PomodoroTaskReportageModel fromMap(Map<String, Object> data) {
        return new PomodoroTaskReportageModel(
                (Integer) data.get(Constants.kTaskReportageId),
                (String) data.get(Constants.kStartDate),
                null,
                (int) data.get(Constants.kStatus),
                (int) data.get(Constants.kPomodoroTaskId),
                (String) data.get(Constants.kTaskName)
        );
    }
}
